action/sealedenvelope.go

```diff
@@ -73,12 +80,23 @@ func (sealed *SealedEnvelope) calcHash() (hash.Hash256, error) {
 		}
 		return rlpSignedHash(tx, sealed.evmNetworkID, sealed.Signature())
 	case iotextypes.Encoding_IOTEX_PROTOBUF:
-		return hash.Hash256b(byteutil.Must(proto.Marshal(sealed.Proto()))), nil
+		return hash.Hash256b(byteutil.Must(sealed.serialize())), nil
 	default:
 		return hash.ZeroHash256, errors.Errorf("unknown encoding type %v", sealed.encoding)
 	}
 }
 
+func (sealed *SealedEnvelope) serialize() ([]byte, error) {
+	buf := bufPool.Get().(*iotextypes.Action)
+	buf.Core = sealed.Envelope.Proto()
+	buf.SenderPubKey = sealed.srcPubkey.Bytes()
+	buf.Signature = sealed.signature
+	buf.Encoding = sealed.encoding
+	ret, err := proto.Marshal(buf)
+	bufPool.Put(buf)
+	return ret, err
+}
```

Comment:
to prevent duplicate logic of deriving `iotextypes.Action`, create a private function
```
func (sealed *SealedEnvelope) proto(act *iotextypes.Action) (
	act.Core = sealed.Envelope.Proto()
	act.SenderPubKey = sealed.srcPubkey.Bytes()
	act.Signature = sealed.signature
	act.Encoding = sealed.encoding
)
```
```
func (sealed *SealedEnvelope) serialize() ([]byte, error) {
	act := bufPool.Get().(*iotextypes.Action)
	sealed.proto(act)
	ret, err := proto.Marshal(act)
	bufPool.Put(act)
	return ret, err
}
```
which will be shared with
https://github.com/iotexproject/iotex-core/pull/3762/files#diff-fda8afdb7165ec30b8d2c40dfd1963def0ffeb117a3d9098167b8139c72527caR124
```
func (sealed *SealedEnvelope) Proto() *iotextypes.Action {
	act := &iotextypes.Action{}
	sealed.proto(act)
	return act
}
```